### ISSUE TITLE
[Bug]: Accessibility Audit of settings screen

### ISSUE THREAD

**Issue by joedolson** [aspirepress/aspireupdate#379]
### What happened?

Hi! This is a light accessibility audit of the AspireUpdate settings screen, requested by @costdev. I'm leaving it all in a single issue, and will leave it to the AspirePress team to break it up as needed.

Accessibility Notes:

- [x] 1) Empty `aria-describedby` target for Enable AspireUpdate API Rewrites.

This won't have any real-world impact on assistive technology, but it's a code smell.

- [x] 2) Skip API rewriting option uses implicit label.

This will make it more difficult for users of voice command tools to control the checkbox. Explicit labeling is a better scenario.

- [x] 3) Skip API rewriting option has a description field that is not associated using `aria-describedby`.

- [x] 4) Generate API Key has no visible text label.

The "Generate API Key" button is visually labeled only using a `title` attribute, which does not appear in many contexts, including touch interfaces and some browser configurations.

- [x] 5) The first few times I tried it, there was no error message exposed visibly or to AT if the API key generation fails due to CORS blocking. However, I later re-tried it and there was an error generated. (Unexpected Error: 0).

The produced error was not associated explicitly with the input field or the button using `aria-describedby`, and was not announced to screen readers using `wp.a11y.speak()`.

Suggest using `wp_get_admin_notice()` to render this error in the AJAX handler for visual consistency with other errors.

I was only able to get this option to generate a CORS error.

- [x] 6) Enable Bug Type group does not have any form of group labeling.

Field groups like checkboxes and radio inputs need a grouping label that provides context for what the group represents. Visually, this is "Enable Debug Type", but that is not associated with the fields.

While the WordPress settings API doesn't support fieldset/legend due to HTML nesting problems, you can use `role="group"` and `aria-labelledby="label"`. Wrap the list of inputs in an element with `role="group" aria-labelledby="target"` and set the target ID on the Enable Debug Type text.

- [x] 7) The text describing debugging ("Outputs the request URL and headers / response...") is not explicitly associated with anything. It doesn't describe a specific field, but rather *all* the above fields. This would be better handled by having better labels on the checkboxes, e.g.

Debug Request URLs
Debug Response Headers
Debug Rewritten request body

This would remove the need for the additional explanation, and provide clearer labeling.

- [x] 8) Voltron easter egg: animation.

The Voltron easter egg has an animation that does not respect `prefers-reduced-animation`. While it's only triggered on `:hover`, it is very large, and the discoverability of what has triggered it is low, due to the 5s delay before it starts. Anybody with both animation-triggered problems and mobility impairments may have a great deal of difficulty ending the animation.

The animation should respect `prefers-reduced-animation`. This does not mean eliminating all animation; it could still animate in, but more slowly and without the repeating animation.

- [x] 9) Voltron easter egg: screen reader experience.

The Voltron ASCII art is readable text with no alternative. Because of this, it's read by screen readers as an arbitrary list of punctuation characters. Knowledgeable screen reader users may figure out that it's a text-based graphic, but will have no way to identify the image. Because graphic is not hidden other than visually, it will always be read by screen reader users, and is not hidden as it is for sighted users.

To keep the easter egg but make it more accessible, mark the `pre` element as `aria-hidden="true"` so that screen reader users don't read it and add an image hidden using `.screen-reader-text` that has an alt text providing the equivalent context, e.g. "Voltron ASCII art easter egg".

- [x] 10) Clear Log & View Log buttons

When there is no log available, these both produce errors. The errors are generated using scripting, and are inserted at the top of the page under the `h1` heading, which will commonly be out of viewport for many users. There is no audible notification using `wp.a11y.speaks()` to let screen reader user know that any action has taken place.

For sighted users, a dynamic notice should appear in proximity to the control that triggered it, for visibility. For screen reader users, there needs to be an audible notification.

- [x] 11) View Log "modal".

"View Log" opens a popup that looks like it should be a modal. It meets some of the parameters for a modal: masks the background content, etc. However, it doesn't meet any of the accessibility requirements. I suggest reviewing the standards for an accessible modal dialog; none of them are met here. The 'close' button is not a button and has no accessible name; the dialog should be using `role="dialog"`; focus should be moved into the dialog; the rest of the page should be marked as `inert`; the dialog should have an accessible name, etc.

- [x] 12) Clear Log

This is OK; using a browser generated dialog to report this will work. However, it's inelegant, because it requires the screen reader to re-announce the page title and a lot of other extraneous text before getting to it. Administering this via an onscreen message and `wp-a11y-speaks` would be better.

- [x] 13) Plugins page notice link text

The link "Learn more" is highly likely to conflict with other link texts produced on this page. I'd suggest changing it to something like "Learn about the Aspire Project" or "Learn about AspireUpdate".

- [x] 14) Plugins page notice

I also suggest that it should be dismissible. I think there should be a persistent indicator, but it does not need to be a full-on notice at all times. Once dismissed, I'd suggest that an adminbar icon would be sufficient.

**General note:**

One overall concern here is the sheer number of different ways of responding to actions. There were notices that were injected at the top of the page using admin notice styling; there were notices injected next to the field that didn't; there was a notice reported using the browser dialog; etc. Consistency is one way of making an interface more accessible, because it is more predictable. For such a simple interface, a user would have to ingest way too many different interactions.

### Code of Conduct

- [x] I agree to follow the Code of Conduct.

**Comment by namithj** [comment]
I think we can actually get rid of the view log model and replace it with a link to the log file opening in a new tab...

Can anyone think of any negative for that approach...

**Comment by costdev** [comment]
Thanks for this @joedolson! I'm starting to action some of the individual items, whereas I think it's best that we resolve the field labels and descriptions in one go, and the same for the Voltron easter egg.

> To keep the easter egg but make it more accessible, mark the pre element as aria-hidden="true" so that screen reader users don't read it and add an image hidden using .screen-reader-text that has an alt text providing the equivalent context, e.g. "Voltron ASCII art easter egg".

Just to check, should the image have an actual `src` too, or is something like this fine?

```html
<pre id="voltron" aria-hidden="true">
<!-- The ASCII Art -->
</pre>
<img src="data:," class="screen-reader-text" alt="Voltron ASCII art easter egg">
```

**Comment by costdev** [comment]
> I think we can actually get rid of the view log model and replace it with a link to the log file opening in a new tab...
> Can anyone think of any negative for that approach...

I don't think there's an issue with just linking to the uniquely-named log file. However, I don't think it's necessary to open the link in a new tab. I think users should generally have control over how links open.

@joedolson Do you have any thoughts on this?

**Comment by joedolson** [comment]
Re: real image.

Realistically, it doesn't even need to be an image; it could be adding `role="img"` to the `pre` element with an `aria-label`. That would probably be the DRYest way of handling it.

Re: link in a new tab. I don't see any good reason to open that in a new tab unless there are un-saved changes in the form; but it's better to have a link that behaves the same way all the time, so toggling that based on whether the form is dirty would be pretty confusing. It can open in a new tab; as long as it's clearly labeled as doing that, it's fine. It just needs to be clear that it will. I prefer links that don't open in new tabs, so that the user can control the link behavior.

**Comment by costdev** [comment]
I'll be going away this weekend, but I have a local branch with some reworking of how we associated option names with the inputs, making error styling more consistent, and placing errors close to the action that triggered them. There may be some additional work required to place _success_ notices close to the action that triggered them, but I'm not sure about that yet.

The local branch should resolve tasks 1, 2, 3, 6, 7, 10, and 12, or at least get us close to resolution with these.

PRs already exist to resolve tasks 4, 8, 9 and 13.

That leaves tasks 5 and 11 that still need to be worked on. 5 may be something I can handle as part of the local branch mentioned above. 11 is a different situation to resolve.

On task 13 (the branding notice), the text change is already in a PR. Next week I can write the code to make this permanently dismissible and open that as an additional PR, but I'm hesitant to do so until we have a clear idea of whether we want to add the proposed admin bar icon, and how it would work. What would its label be? Would it be a visible label, or an `aria-label`? Where would we place it in the admin bar? etc. - I'd welcome thoughts on this so we can move this forward.

**Comment by joedolson** [comment]
Suggestion for the icon:

Two states: "AspirePress Enabled", "AspirePress Disabled", as a button with a popover that reveals state & different visible states representing that. A properly structured popover is essentially self-labeling.

For location, I think the admin menu end. It's a status condition, and *essentially* so is the profile menu; they both communication about the current state of the admin session, which makes it a reasonable grouping.

**Comment by costdev** [comment]
Task 14 is now complete. I'll open a PR for task 11 tomorrow.

**Comment by costdev** [comment]
PR #397 takes on task 11. It needs some accessibility feedback before we can consider it for merge.

**Comment by costdev** [comment]
Noting for the future: We should look at creating an accessible modal to retain the single-page experience that exists for the other features on the settings page. When I get some time, I'll look into this more and open an issue.
